File: family-os/src/agents/triage.ts
import type { Signal, Urgency } from "../core/types.ts";
import { relativeDays, type Clock, type LocalDateTime } from "../core/time.ts";

export interface Triage {
  urgency: Urgency;
  dueAt?: LocalDateTime;
  reasons: string[];
}

/** Decide how loudly a signal should reach a parent, and by when it must be resolved. */
export function triage(signal: Signal, clock: Clock): Triage {
  const now = clock.now();
  const ex = signal.extracted;
  const reasons: string[] = [];
  if (signal.kind === "fyi") return { urgency: "fyi", reasons: ["Nothing is asked of you."] };

  const dueAt = ex.deadline ?? (signal.kind === "schedule_change" || signal.kind === "coparent_message" || signal.kind === "appointment" ? ex.when?.start : undefined) ?? ex.when?.start;
  const daysToDue = dueAt ? relativeDays(now, dueAt) : undefined;
  const daysToEvent = ex.when ? relativeDays(now, ex.when.start) : undefined;

  if (ex.deadline) reasons.push(`Deadline in ${daysToDue} day${daysToDue === 1 ? "" : "s"}.`);
  if (ex.when) reasons.push(`Happens in ${daysToEvent} day${daysToEvent === 1 ? "" : "s"}.`);
  if (ex.tone?.hostile) reasons.push("Tone flagged; handled through the neutral channel.");

  let urgency: Urgency;
  if (daysToDue !== undefined && daysToDue <= 3) urgency = "now";
  else if (signal.kind === "schedule_change" && daysToEvent !== undefined && daysToEvent <= 7) urgency = "now";
  else if (daysToDue !== undefined && daysToDue <= 10) urgency = "this_week";
  else if (daysToEvent !== undefined && daysToEvent <= 10) urgency = "this_week";
  else urgency = "later";
  if (!ex.when && !ex.deadline) {
    urgency = signal.kind === "coparent_message" ? "now" : "this_week";
    reasons.push("Undated; surfacing so it is not lost.");
  }
  return { urgency, dueAt, reasons };
}

File: src/main/java/app/familyos/service/application/pipeline/Triage.java
package app.familyos.service.application.pipeline;

import app.familyos.service.domain.proposals.Urgency;
import app.familyos.service.domain.signals.Extracted;
import app.familyos.service.domain.signals.Signal;
import app.familyos.service.domain.signals.SignalKind;
import app.familyos.service.infrastructure.time.Times;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/** How loudly a signal should reach a parent, and by when it must be resolved. */
public final class Triage {
    private Triage() {
    }

    public record Result(Urgency urgency, LocalDateTime dueAt, List<String> reasons) {
    }

    public static Result triage(Signal signal, LocalDateTime now) {
        Extracted ex = signal.getExtracted();
        List<String> reasons = new ArrayList<>();
        if (signal.getKind() == SignalKind.FYI) return new Result(Urgency.FYI, null, List.of("Nothing is asked of you."));
        LocalDateTime dueAt = ex.getDeadline() != null ? ex.getDeadline() : ex.getWhen() != null ? ex.getWhen().start() : null;
        Long daysToDue = dueAt == null ? null : Times.daysBetween(now, dueAt);
        Long daysToEvent = ex.getWhen() == null ? null : Times.daysBetween(now, ex.getWhen().start());
        if (ex.getDeadline() != null) reasons.add("Deadline in " + daysToDue + " day" + (daysToDue == 1 ? "" : "s") + ".");
        if (ex.getWhen() != null) reasons.add("Happens in " + daysToEvent + " day" + (daysToEvent == 1 ? "" : "s") + ".");
        if (ex.getTone() != null && ex.getTone().hostile()) reasons.add("Tone flagged; handled through the neutral channel.");
        Urgency urgency;
        if (daysToDue != null && daysToDue <= 3) urgency = Urgency.NOW;
        else if (signal.getKind() == SignalKind.SCHEDULE_CHANGE && daysToEvent != null && daysToEvent <= 7) urgency = Urgency.NOW;
        else if (daysToDue != null && daysToDue <= 10) urgency = Urgency.THIS_WEEK;
        else if (daysToEvent != null && daysToEvent <= 10) urgency = Urgency.THIS_WEEK;
        else urgency = Urgency.LATER;
        if (ex.getWhen() == null && ex.getDeadline() == null) {
            urgency = signal.getKind() == SignalKind.COPARENT_MESSAGE ? Urgency.NOW : Urgency.THIS_WEEK;
            reasons.add("Undated; surfacing so it is not lost.");
        }
        return new Result(urgency, dueAt, reasons);
    }
}
